Run CMS locally with supporting services in Docker (#38)
ONSdigital/dis-wagtail: README.md

Follow these steps to set up and run the project using Docker.

1. **Build and Start the Containers**

    ```bash
    # pull the supporting containers
    make compose-pull

    # build the main application image

>     make docker-build

Reviewer: ```suggestion
    make compose-build
```

    # start the containers
    make compose-up
    ```

2. **Migrations and Superuser Creation**

    If this is your first time setting up the project, you’ll need to run migrations to set up the database schema and create an administrative user.

    ```bash
    # ssh into the web container
    make docker-shell

    # Run database migrations
    make migrate

    # Create a superuser for accessing the admin interface
    make createsuperuser
    ```

3. **Start Django Inside the Container**

    Once the containers are running, you need to manually start Django from within the web container. This allows for running both the Django server and any additional background services (e.g., schedulers).

> [!WARNING]  

> > The `honcho` command will pick up your local mounted `.env` file when running via `docker-compose`. Ensure that you comment out any local variables which might cause clashes in the container context as they will take precedence when running `honcho start`.

Reviewer: This breaks the indentation and doesn't escape the markdown. I suggest just using a warning emoji, as it doesn't look like you can use indented panels.
<img width="129" alt="image" src="https://github.com/user-attachments/assets/68241581-41b1-403c-92b8-7c0568a92e41">

    ```bash
    # Start both Django and the scheduler using Honcho
    honcho start

    # This is not needed if you used `honcho start`
    make runserver
    ```

You can then access the admin at `http://0.0.0.0:8000/admin/` or `http://localhost:8000/admin/`.

#### Running locally with supporting services in Docker

You can also run the main application locally with the supporting backend services such as the Postgres and Redis running in Docker. This can be useful when you want to make changes that require the app to be frequently restarted to be picked up.

In order to run it:

1. Pull the images of the supporting services.

```
make compose-dev-pull
```

2. Start the supporting services in Docker.

```
make compose-dev-up
```

3. Run the below command to run the necessary pre-run steps:

```

> make dev-init

Reviewer: Either `load-design-system-templates` needs to be added to `dev-init` or needs to be added as a new step.

Author: Added to `dev-init`

> make dev-init

Reviewer: This should mention that an admin user will be created with user `admin` and pass `changeme`; otherwise, they won't know unless they dig around.

```

4. Run the Django server locally via your IDE or with the following command:

```

> make runserver

Reviewer 2: The current README has `dj runserver 0.0.0.0:8000` so should it not be `make runserver 0.0.0.0:8000` ? I don't suppose this will have a large effect but thought it be better to point out.

Reviewer 3: The default is 127.0.0.1:8000 [ref](https://docs.djangoproject.com/en/5.1/ref/django-admin/#runserver), so this is fine

```

> You can specify the runtime configuration either in your IDE (for PyCharm see [here](https://www.jetbrains.com/help/pycharm/run-debug-configuration.html#createExplicitly)), or copy the `.env-dev-example` and rename it to `.env` which will allow Django to pick up the config.

Reviewer: `.env-dev-example` has been renamed to `.development.env`

Reviewer: This is still outstanding.

Author: I swear I've changed that 🤔 Done now!

> Note that once you create the `.env` file, and you'd like to switch back to running the application in a container with `make compose-up`, the `.env` file will be accessible inside the containers and it will be picked up by the `honcho` command. In order to avoid conflicts you should comment out the `DATABASE_URL` and `REDIS_URL` variables in the `.env` file.

Reviewer 4: ```suggestion
Note that once you create the `.env` file, and you'd like to switch back to running the application in a container with `make compose-up`, the `.env` file will be accessible inside the containers and it will be picked up by the `honcho` command. In order to avoid conflicts you may need to comment out some variables (such as `DATABASE_URL` and `REDIS_URL`).
```

> [!NOTE]

> > When running the application in a virtual environment via Poetry the `.env` file will not be picked up automatically. For this to work you'll need to install the [poetry-plugin-dotenv](https://github.com/pivoshenko/poetry-plugin-dotenv). However if you installed Poetry with `brew` rather than `pip` that currently isn't going to work (see the [issue](https://github.com/pivoshenko/poetry-plugin-dotenv/issues/327)) and you'll need to install an older and seemingly no longer maintained [poetry-dotenv-plugin](https://github.com/mpeteuil/poetry-dotenv-plugin).

Reviewer: We should move both of these up as they are required before the app can run. We should inform the users of anything they need before they need it. These are essentially prerequisites.

Follow these steps to set up and run the project using Docker.

1. **Build and Start the Containers**

    ```bash
    # pull the supporting containers
    make compose-pull

    # build the main application image
    make compose-build

    # start the containers
    make compose-up
    ```

2. **Migrations and Superuser Creation**

    If this is your first time setting up the project, you’ll need to run migrations to set up the database schema and create an administrative user.

    ```bash
    # ssh into the web container
    make docker-shell

    # Run database migrations
    make migrate

    # Create a superuser for accessing the admin interface
    make createsuperuser
    ```

3. **Start Django Inside the Container**

    Once the containers are running, you need to manually start Django from within the web container. This allows for running both the Django server and any additional background services (e.g., schedulers).

>     ⚠️ WARNING  

Reviewer: You can still use the panel.
```suggestion
    > ⚠️ WARNING  
```

Author: Done, `make format` enforces that we put a chevron before the text on line 123 as well

    # To run just the web server. alias: djrun
    # This is not needed if you used `honcho start`
    make runserver
    ```

You can then access the admin at `http://0.0.0.0:8000/admin/` or `http://localhost:8000/admin/`.

#### Running slocally with supporting services in Docker

You can also run the main application locally with the supporting backend services such as the Postgres and Redis running in Docker. This can be useful when you want to make changes that require the app to be frequently restarted to be picked up.

In order to run it:

1. Pull the images of the supporting services.

```
make compose-dev-pull
```

2. Start the supporting services in Docker.

```
make compose-dev-up
```

3. Run the following pre-run steps:

```
make makemigrations
make collectstatic
make migrate

> make createsuperuser

Author: Yeah, another thing I wanted to add, but I was concerned that the Makefile is getting bloated 😄 

If we're adding this, what do you think about also adding a makefile recipe for the [one liner](https://github.com.mcas.ms/torchbox/wagtailmedia/blob/main/tox.ini#L77) to create an admin user that you've linked at the tech session last week? That would make the `make dev-init` command even easier to run as it won't stop waiting for input.

> make createsuperuser

Reviewer 3: we spoke in Slack. yes to adding it to `make dev-init`

> make createsuperuser

Author: We've talked about it and I'll add it add it as part of `make dev-init` rather than adding a new `make` recipe.

You can then access the admin at `http://0.0.0.0:8000/admin/` or `http://localhost:8000/admin/`.

> #### Running Wagtail locally with supporting services in Docker

Reviewer 5: Add this to table of contents please =)

Author: 🫡

You can also run the main application locally with the supporting backend services such as the Postgres and Redis running in Docker. This can be useful when you want to make changes that require the app to be frequently restarted to be picked up.

> In order tu run it:

Reviewer 2: to instead of tu?

1. Pull the images of the supporting services.

```
make docker-dev-pull
```

2. Start the supporting services in Docker.

```
make docker-dev-up
```

3. Run the following pre-run steps:

```
make makemigrations
make collectstatic
make migrate
make createsuperuser
```

4. Run the Django server locally via your IDE or with the following command:

```
make runserver
```

You can specify the runtime configuration either in your IDE (for PyCharm see [here](https://www.jetbrains.com/help/pycharm/run-debug-configuration.html#createExplicitly)), or copy the `.env-dev-example` and rename it to `.env` which will allow Django to pick up the config.

> [!NOTE]

> > When running the application in a virtual environment via Poetry the `.env` file will not be picked up automatically. For this to work you'll need to install the [poetry-plugin-dotenv](https://github.com/pivoshenko/poetry-plugin-dotenv). However if you installed Poetry with `brew` rather than `pip` that currently isn't going to work (see the [issue](https://github.com/pivoshenko/poetry-plugin-dotenv/issues/327)) and you'll need to install an older and seemingly no longer maintained [poetry-dotenv-plugin](https://github.com/mpeteuil/poetry-dotenv-plugin).

Reviewer 5: Nit: This maybe helpful for other devs. Maybe you can add a line about this in the `README` file.

I have installed Poetry using [the official installer](https://python-poetry.org/docs/#installing-with-the-official-installer). I think the recommended way is Pipx or using the official installer.

Not sure installing Poetry using pip is a good idea. https://github.com/orgs/python-poetry/discussions/8687

Since I used the official installer, I didn't have to do any of the steps in the `README` =)

Reviewer 5: Also wouldn't this be a dev dependency? If we truly need this, which I didn't need due to installing it using the official installer.

```bash
# Check if you need to generate any new migrations after changes to the model
make check-makemigrations

# Generate migrations
make makemigrations

# Apply migrations. Needed if new migrations have been generated (either by you, or via upstream code)

> make migrate

Reviewer: I suggest replacing all the `dj` alias commands like `djrun` to use the makefile target instead so we have a consistent commands to run regardless of whether we are running via docker or locally.

Reviewer 5: I really don't have a strong preference haha.

I mean I agree with the approach that @MebinAbraham is suggesting but when it comes to removing them from `/.docker/bashrc.sh` it makes to get rid of it, unless there is a solid reason to keep those still.

Reviewer 3: The `dj` aliases are something we've been using at Torchbox as a shortcut, so not wedded to it
